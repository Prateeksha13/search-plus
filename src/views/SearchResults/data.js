export const products = {
  "iPhone 13": {
    id: "p1",
    name: "iPhone 13",
    models: [
      {
        id: "p1m1",
        name: "iPhone 13",
        finish: ["Green", "Pink", "Blue", "Midnight", "Starlight", "Red"],
        storage: ["128 GB", "256 GB", "512 GB"],
        display: "",
      },
      {
        id: "p1m2",
        name: "iPhone 13 Mini",
        finish: ["Green", "Pink", "Blue", "Midnight", "Starlight", "Red"],
        storage: ["128 GB", "256 GB", "512 GB"],
      },
    ],
  },
  "iPhone 13 Pro": {
    id: "p2",
    name: "iPhone 13 Pro",
    models: [
      {
        id: "p2m1",
        name: "iPhone 13 Pro",
        finish: ["Alpine Green", "Silver", "Gold", "Graphite", "Sierra Blue"],
        storage: ["128 GB", "256 GB", "512 GB", "1 TB"],
        display: "6.1 inch",
      },
      {
        id: "p2m2",
        name: "iPhone 13 Pro Max",
        finish: ["Green", "Silver", "Gold", "Graphite", "Sierra Blue"],
        storage: ["128 GB", "256 GB", "512 GB", "1 TB"],
        display: "6.7 inch",
      },
    ],
  },
};

export const locations = [
  {
    name: "United States",
    type: "country",
    query: "location:united%20states&store=",
  },
  {
    name: "Ohio, United States",
    country: "United States",
    type: "administrative_area",
  },
  {
    name: "California, United States",
    country: "United States",
    type: "administrative_area",
  },
];

export const stores = [
  {
    id: "r001",
    name: "Crocker Park",
    address: [
      "267 Crocker Park Boulevard, Westlake, OH 44145",
      "(440) 788-2800",
    ],
    administrative_area: "Ohio, United States",
    locality: "Westlake",
    country: "United States",
  },
  {
    id: "r002",
    name: "Eastern Town Center",
    address: ["4210 The Strand, Columbus, OH 43219", "(614) 934-2810"],
    administrative_area: "Ohio, United States",
    locality: "Columbus",
    country: "United States",
  },
  {
    id: "r003",
    name: "Eton",
    address: ["28849 Chagrin Boulevard, Woodmere, OH 44122", "(216) 535-4560"],
    administrative_area: "Ohio, United States",
    locality: "Woodmere",
    country: "United States",
  },
  {
    id: "r004",
    name: "Franklin Park Mall",
    address: ["5001 Monroe Street, Toledo, OH 43623", "(419) 473-4250"],
    administrative_area: "Ohio, United States",
    locality: "Toledo",
    country: "United States",
  },
];

export const inventory = [
  {
    id: "inventory0001",
    product: "iPhone 13 Pro",
    store_id: "r001",
    model: "iPhone 13 Pro",
    finish: "Alpine Green",
    storage: "256 GB",
    stock: 2,
  },
  {
    id: "inventory0002",
    product: "iPhone 13 Pro",
    store_id: "r001",
    model: "iPhone 13 Pro",
    finish: "Alpine Green",
    storage: "512 GB",
    stock: 5,
  },
  {
    id: "inventory0003",
    product: "iPhone 13",
    store_id: "r001",
    model: "iPhone 13",
    finish: "Pink",
    storage: "512 GB",
    stock: 4,
  },
  {
    id: "inventory0004",
    product: "iPhone 13 Pro",
    store_id: "r002",
    model: "iPhone 13 Pro",
    finish: "Alpine Green",
    storage: "512 GB",
    stock: 15,
  },
  {
    id: "inventory0005",
    product: "iPhone 13",
    store_id: "r001",
    model: "iPhone 13 Mini",
    finish: "Green",
    storage: "128 GB",
    stock: 4,
  },
];
