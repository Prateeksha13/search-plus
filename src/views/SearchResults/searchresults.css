.home-search-bar-background {
  background-color: var(--backgroud-light);
  padding: 16px;
  margin: 150px 164px 45px;
  border-radius: 15px;
}

.search-bar-background {
  background-color: var(--backgroud-light);
  padding: 16px 172px;
}

.search-result-list {
  margin: 0px 150px;
  padding: 64px 22px 45px;
}

.search-filters {
  display: flex;
  padding: 16px 164px;
  border-bottom: 1px solid var(--border-dark);
  background-color: var(--white);
}

.search-filters .dropdown_wrapper {
  min-width: 200px;
  margin: 0 8px;
  flex: 1;
}

.top-sticky-controls {
  position: sticky;
  top: 44px;
}
