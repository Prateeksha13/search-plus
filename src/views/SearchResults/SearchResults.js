import React, { useEffect, useState } from "react";

import Layout from "../../components/Layout/Layout";
import SearchBar from "../../components/SearchBar/SearchBar";
import SearchResultItem from "../../components/SearchResultItem/SearchResultItem";
import Select from "../../components/Select/Select";
import { products, locations, stores, inventory } from "./data";
import "./searchresults.css";

const SearchResults = ({}) => {
  const [selectedProduct, setSelectedProduct] = useState({ name: "" });
  const [selectedModel, setSelectedModel] = useState({ name: "" });
  const [selectedLocation, setSelectedLocation] = useState("");
  const [searchResponse, setSearchResponse] = useState({});
  const [selectedFinish, setSelectedFinish] = useState("Any");
  const [selectedStorage, setSelectedStorage] = useState("Any");
  const [selectedAvailability, setSelectedAvailability] = useState("Yes/No");
  const [showFilters, setShowFilters] = useState(false);
  const [showNoDataMessage, setShowNoDataMessage] = useState(false);

  useEffect(() => {
    onFilter();
  }, [selectedModel, selectedFinish, selectedStorage, selectedAvailability]);

  const getProductSuggestions = (value) => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;
    let filteredValues = Object.keys(products)
      .filter(
        (key) =>
          products[key].name.toLowerCase().slice(0, inputLength) === inputValue
      )
      .map((key) => ({ ...products[key], label: products[key].name }));

    return inputLength === 0 ? [] : filteredValues;
  };

  const getProductSuggestionValue = (suggestion) => suggestion.name;

  const renderProductSuggestions = (suggestion) => (
    <div>{suggestion.label}</div>
  );

  const informProductSuggestionSelect = (suggestion) => {
    setSelectedProduct(suggestion);
  };

  const getLocationSuggestions = (value) => {
    const inputValue = value.trim().toLowerCase();
    const inputLength = inputValue.length;
    let filteredValues = locations
      .filter(
        (location) =>
          `${location.name} ${location.country}`
            .toLowerCase()
            .slice(0, inputLength) === inputValue
      )
      .map((location) => ({
        ...location,
        label: location.name,
      }));

    return inputLength === 0 ? [] : filteredValues;
  };

  const getLocationSuggestionValue = (suggestion) => suggestion.name;

  const renderLocationSuggestion = (suggestion) => (
    <div>{suggestion.label}</div>
  );

  const informLocationSuggestionSelect = (suggestion, suggestionValue) => {
    setSelectedLocation(suggestion);
    if (selectedLocation.name) {
      onSearch();
    }
    console.log(suggestion, suggestionValue);
  };

  const onSearch = () => {
    setShowFilters(true);
    setShowNoDataMessage(true);
    if (!selectedLocation) {
      setSelectedLocation({
        name: "United States",
        type: "country",
        query: "location:united%20states&store=",
      });
    }
    setSelectedModel(
      selectedProduct.models && selectedProduct.models[0]
        ? selectedProduct.models[0]
        : { name: "" }
    );
    let stockData = {};
    stores.forEach((store) => {
      if (
        store[selectedLocation.type] &&
        store[selectedLocation.type] === selectedLocation.name
      ) {
        stockData[store.id] = {
          store: store,
          stock: 0,
        };
      }
    });

    let response = {};
    inventory.forEach((item) => {
      if (
        item.product === selectedProduct.name &&
        item.model === selectedModel.name &&
        (item.finish === selectedFinish || selectedFinish === "Any") &&
        (item.storage === selectedStorage || selectedStorage === "Any") &&
        Object.keys(stockData).indexOf(item.store_id) > -1
      ) {
        stockData[item.store_id] = {
          ...stockData[item.store_id],
          stock: stockData[item.store_id].stock + item.stock,
        };
      }
    });

    if (selectedAvailability === "Yes") {
      Object.keys(stockData).forEach((storeId) => {
        if (stockData[storeId].stock > 0) {
          response[storeId] = stockData[storeId];
        }
      });
    } else if (selectedAvailability === "No") {
      Object.keys(stockData).forEach((storeId) => {
        if (stockData[storeId].stock === 0) {
          response[storeId] = stockData[storeId];
        }
      });
    } else {
      response = { ...stockData };
    }

    setSearchResponse(response);
  };

  const onFilter = () => {
    let stockData = {};
    stores.forEach((store) => {
      if (
        store[selectedLocation.type] &&
        store[selectedLocation.type] === selectedLocation.name
      ) {
        stockData[store.id] = {
          store: store,
          stock: 0,
        };
      }
    });

    let response = {};
    inventory.forEach((item) => {
      if (
        item.model === selectedModel.name &&
        (item.finish === selectedFinish || selectedFinish === "Any") &&
        (item.storage === selectedStorage || selectedStorage === "Any") &&
        Object.keys(stockData).indexOf(item.store_id) > -1
      ) {
        stockData[item.store_id] = {
          ...stockData[item.store_id],
          stock: stockData[item.store_id].stock + item.stock,
        };
      }
    });

    if (selectedAvailability === "Yes") {
      Object.keys(stockData).forEach((storeId) => {
        if (stockData[storeId].stock > 0) {
          response[storeId] = stockData[storeId];
        }
      });
    } else if (selectedAvailability === "No") {
      Object.keys(stockData).forEach((storeId) => {
        console.log(stockData[storeId].stock, "no");
        if (stockData[storeId].stock === 0) {
          response[storeId] = stockData[storeId];
        }
      });
    } else {
      response = { ...stockData };
    }
    setSearchResponse(response);
  };

  const onModelChange = (value) => {
    setSelectedModel(value.value);
  };

  const onFinishChange = (value) => {
    setSelectedFinish(value.value);
  };

  const onStorageChange = (value) => {
    setSelectedStorage(value.value);
  };

  const onAvailabilityChange = (value) => {
    setSelectedAvailability(value.value);
  };

  return (
    <Layout>
      <div className="search-result-content">
        <div className="top-sticky-controls">
          <div className="search-bar-background">
            <SearchBar
              onSearch={onSearch}
              getProductSuggestions={getProductSuggestions}
              renderProductSuggestions={renderProductSuggestions}
              informProductSuggestionSelect={informProductSuggestionSelect}
              getProductSuggestionValue={getProductSuggestionValue}
              getLocationSuggestions={getLocationSuggestions}
              renderLocationSuggestion={renderLocationSuggestion}
              informLocationSuggestionSelect={informLocationSuggestionSelect}
              getLocationSuggestionValue={getLocationSuggestionValue}
            />
          </div>
          {selectedProduct.name && showFilters && (
            <div className="search-filters">
              <Select
                data={
                  "models" in selectedProduct &&
                  selectedProduct.models.map((model) => ({
                    label: model.name,
                    value: model,
                  }))
                }
                value={{ label: selectedModel.name, value: selectedModel }}
                onChange={onModelChange}
              />
              <Select
                data={[{ label: "Any", value: "Any" }].concat(
                  selectedModel &&
                    "finish" in selectedModel &&
                    selectedModel.finish.map((item) => ({
                      label: item,
                      value: item,
                    }))
                )}
                value={{ label: selectedFinish, value: selectedFinish }}
                onChange={onFinishChange}
              />
              <Select
                data={[{ label: "Any", value: "Any" }].concat(
                  selectedModel &&
                    "storage" in selectedModel &&
                    selectedModel.storage.map((item) => ({
                      label: item,
                      value: item,
                    }))
                )}
                value={{ label: selectedStorage, value: selectedStorage }}
                onChange={onStorageChange}
              />
              <Select
                data={[
                  {
                    label: "Yes/No",
                    value: "Yes/No",
                  },
                  {
                    label: "Yes",
                    value: "Yes",
                  },
                  {
                    label: "No",
                    value: "No",
                  },
                ]}
                value={{
                  label: selectedAvailability,
                  value: selectedAvailability,
                }}
                onChange={onAvailabilityChange}
              />
            </div>
          )}
        </div>
        <div className="search-result-list">
          {Object.keys(searchResponse).map((storeId, index) => (
            <SearchResultItem
              key={storeId}
              itemNumber={index + 1}
              locality={searchResponse[storeId].store.locality}
              storeName={searchResponse[storeId].store.name}
              storeAddress={searchResponse[storeId].store.address}
              isProductAvailable={
                searchResponse[storeId].stock > 0 ? true : false
              }
              stockQty={searchResponse[storeId].stock}
            />
          ))}
          {Object.keys(searchResponse).length === 0 && showNoDataMessage && (
            <center>
              <h4>No results</h4>
            </center>
          )}
        </div>
      </div>
    </Layout>
  );
};

export default SearchResults;
