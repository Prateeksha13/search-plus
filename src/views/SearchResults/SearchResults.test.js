import React from "react";
import renderer from "react-test-renderer";
import SearchResults from "./SearchResults";

test("renders search results", () => {
  const component = renderer.create(<SearchResults />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
