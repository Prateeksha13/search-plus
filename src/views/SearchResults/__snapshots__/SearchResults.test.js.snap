// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders search results 1`] = `
<div
  className="app-page"
>
  <header
    className="app-header"
    role="header"
  >
    <div
      className="app-header-content"
    >
      <a
        className="app-header-brand"
        href="/"
      >
        <img
          alt="Apple"
          src="apple-logo.svg"
        />
        <span
          className="app-header-brand-name"
        >
          Search+
        </span>
      </a>
      <div />
    </div>
  </header>
  <div
    className="app-page-content"
  >
    <div
      className="search-result-content"
    >
      <div
        className="top-sticky-controls"
      >
        <div
          className="search-bar-background"
        >
          <div
            className="search-bar-container"
          >
            <div
              className="search-bar-input-wrapper"
            >
              <div
                className="search-field-wrapper"
              >
                <img
                  alt="search"
                  className="search-field-icon"
                  src="search.svg"
                />
                <div
                  aria-expanded={false}
                  aria-haspopup="listbox"
                  aria-owns="react-autowhatever-1"
                  className="react-autosuggest__container"
                  role="combobox"
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls="react-autowhatever-1"
                    autoComplete="off"
                    autoFocus={true}
                    className="react-autosuggest__input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Search by product name"
                    type="text"
                    value=""
                  />
                  <div
                    className="react-autosuggest__suggestions-container"
                    id="react-autowhatever-1"
                    role="listbox"
                  />
                </div>
              </div>
              <span
                className="horizontal-divider"
              />
              <div
                className="search-field-wrapper"
              >
                <img
                  alt="search"
                  className="search-field-icon"
                  src="search.svg"
                />
                <div
                  aria-expanded={false}
                  aria-haspopup="listbox"
                  aria-owns="react-autowhatever-1"
                  className="react-autosuggest__container"
                  role="combobox"
                >
                  <input
                    aria-activedescendant={null}
                    aria-autocomplete="list"
                    aria-controls="react-autowhatever-1"
                    autoComplete="off"
                    className="react-autosuggest__input"
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    placeholder="Search by location name"
                    type="text"
                    value=""
                  />
                  <div
                    className="react-autosuggest__suggestions-container"
                    id="react-autowhatever-1"
                    role="listbox"
                  />
                </div>
              </div>
            </div>
            <button
              className="custom-button button-primary"
              onClick={[Function]}
              role="button"
              type="button"
            >
              Search
            </button>
          </div>
        </div>
      </div>
      <div
        className="search-result-list"
      />
    </div>
  </div>
</div>
`;
